forastero/scoreboard.py

```diff
@@ -203,8 +213,28 @@ async def _dequeue(self) -> tuple[BaseTransaction, BaseTransaction]:
         await self._q_mon.wait_for_not_empty()
         # When a monitor transaction arrives, lock out the drain procedure
         await self._lock.acquire()
-        # Wait for the reference model to provide a transaction
-        next_ref = await self._q_ref.pop()
+        # If matching window is 1, wait for the first reference transaction
+        if self.match_window == 1:
+            next_ref = await self._q_ref.pop()
+        # If matching window > 1, peek at the captured transaction and search
+        # ahead the next N transactions
+        else:
+            peek_mon = self._q_mon[0]
+            next_ref = None
+            while next_ref is None:
+                # Search within the match window
+                for idx, peek_ref in enumerate(self._q_ref[: self.match_window]):
+                    if peek_ref == peek_mon:
+                        next_ref = await self._q_ref.pop(idx)
+                        break
+                # If no match found...
+                else:
+                    # ...if queue length less than the match window, wait for a push
+                    if len(self._q_ref) < self.match_window:
+                        await self._q_ref.on_push_event.wait()
+                    # ...otherwise pop the zeroeth entry
+                    else:
+                        next_ref = await self._q_ref.pop(0)
```

Comment:
If there is no match and you return the head of the queue (pop from queue), which will mismatch, will a later transaction that matched the popped entry also mismatch as the entry is no longer in queue and pop the head and so on? If so, the TB (if not in fail fast mode) could report more errors than there are. The first mismatch is the one that you would debug so I guess it doesn’t really matter.